// @0gfoundation/0g-storage-ts-sdk/lib.esm/transfer/types.js
export const defaultUploadOption = {
    tags: '0x',
    submitter: '',
    finalityRequired: true,
    taskSize: 1,
    expectedReplica: 1,
    fragmentSize: 1024 * 1024 * 1024 * 4, // 4GB
    skipTx: false,
    skipIfFinalized: true,
    fee: BigInt(0),
};
/**
 * Merges user-provided upload options with default values
 */
export function mergeUploadOptions(userOptions = {}) {
    return {
        tags: userOptions.tags ?? defaultUploadOption.tags,
        submitter: userOptions.submitter ?? defaultUploadOption.submitter,
        finalityRequired: userOptions.finalityRequired ??
            defaultUploadOption.finalityRequired,
        taskSize: userOptions.taskSize ?? defaultUploadOption.taskSize,
        expectedReplica: userOptions.expectedReplica ?? defaultUploadOption.expectedReplica,
        fragmentSize: userOptions.fragmentSize ?? defaultUploadOption.fragmentSize,
        skipTx: userOptions.skipTx ?? defaultUploadOption.skipTx,
        skipIfFinalized: userOptions.skipIfFinalized ?? defaultUploadOption.skipIfFinalized,
        fee: userOptions.fee ?? defaultUploadOption.fee,
        nonce: userOptions.nonce,
        onProgress: userOptions.onProgress,
        encryption: userOptions.encryption,
    };
}
