// @0gfoundation/0g-storage-ts-sdk/lib.commonjs/file/EncryptedFile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EncryptedFileFragment = exports.EncryptedFile = void 0;
exports.newSymmetricEncryptedFile = newSymmetricEncryptedFile;
exports.newEciesEncryptedFile = newEciesEncryptedFile;
const AbstractFile_js_1 = require("./AbstractFile.js");
const index_js_1 = require("./Iterator/index.js");
const utils_js_1 = require("./utils.js");
const encryption_js_1 = require("../common/encryption.js");
// EncryptedFile wraps any AbstractFile with AES-256-CTR encryption. It
// prepends the header (17 bytes for v1, 50 bytes for v2) to the data stream
// and encrypts inner bytes on the fly when `readFromFile` is called. This
// is transparent to the Merkle-tree and segment pipelines.
class EncryptedFile extends AbstractFile_js_1.AbstractFile {
    inner;
    key;
    header;
    constructor(inner, key, header) {
        super();
        this.inner = inner;
        this.key = key;
        this.header = header;
        this.offset = 0;
        this.size_ = inner.size() + header.size();
        this.paddedSize_ = (0, utils_js_1.iteratorPaddedSize)(this.size_, true);
    }
    createFragment(offset, size, _paddedSize) {
        return new EncryptedFileFragment(this, offset, size);
    }
    async readFromFile(start, end) {
        if (start < 0 || start >= this.size() || start >= end) {
            throw new Error('invalid start offset');
        }
        if (end > this.size())
            end = this.size();
        const total = end - start;
        const buf = new Uint8Array(total);
        const headerSize = this.header.size();
        let written = 0;
        // Copy any header bytes that fall within [start, end).
        if (start < headerSize) {
            const headerBytes = this.header.toBytes();
            const from = start;
            const to = Math.min(headerSize, end);
            buf.set(headerBytes.slice(from, to), 0);
            written += to - from;
        }
        // Copy encrypted inner data for the remainder.
        if (written < total) {
            const innerStart = start < headerSize ? 0 : start - headerSize;
            const innerEnd = end - headerSize;
            if (innerEnd > innerStart) {
                const inner = await this.inner.readFromFile(innerStart, innerEnd);
                const innerBuf = new Uint8Array(inner.buffer);
                (0, encryption_js_1.cryptAt)(this.key, this.header.nonce, innerStart, innerBuf);
                buf.set(innerBuf, written);
                written += innerBuf.length;
            }
        }
        return { bytesRead: written, buffer: buf };
    }
    iterateWithOffsetAndBatch(offset, batch, flowPadding) {
        // MemIterator pulls bytes via this.file.readFromFile
        // (src.ts/file/Iterator/MemIterator.ts line 66) — EncryptedFile
        // overrides readFromFile, so MemIterator works unchanged.
        const paddedSize = (0, utils_js_1.iteratorPaddedSize)(this.size(), flowPadding);
        return new index_js_1.MemIterator(this, offset, batch, paddedSize);
    }
}
exports.EncryptedFile = EncryptedFile;
// Fragment of an EncryptedFile — delegates readFromFile to the parent with
// an offset adjustment.
class EncryptedFileFragment extends AbstractFile_js_1.AbstractFile {
    parent;
    constructor(parent, offset, size) {
        super();
        this.parent = parent;
        this.offset = offset;
        this.size_ = size;
        this.paddedSize_ = (0, utils_js_1.iteratorPaddedSize)(size, true);
    }
    createFragment(_offset, _size, _paddedSize) {
        // Fragments are not further splittable; mirror the Go reference.
        return this;
    }
    async readFromFile(start, end) {
        if (start < 0 || start >= this.size() || start >= end) {
            throw new Error('invalid start offset');
        }
        if (end > this.size())
            end = this.size();
        return this.parent.readFromFile(this.offset + start, this.offset + end);
    }
    iterateWithOffsetAndBatch(offset, batch, flowPadding) {
        const paddedSize = (0, utils_js_1.iteratorPaddedSize)(this.size(), flowPadding);
        return new index_js_1.MemIterator(this, offset, batch, paddedSize);
    }
}
exports.EncryptedFileFragment = EncryptedFileFragment;
// Convenience constructors matching the Go sdk's NewEncryptedData /
// NewEncryptedDataECIES.
function newSymmetricEncryptedFile(inner, key) {
    if (key.length !== 32)
        throw new Error('key must be 32 bytes');
    return new EncryptedFile(inner, key, (0, encryption_js_1.newSymmetricHeader)());
}
function newEciesEncryptedFile(inner, recipientPub) {
    const { header, key } = (0, encryption_js_1.newEciesHeader)(recipientPub);
    return new EncryptedFile(inner, key, header);
}
