// @0gfoundation/0g-storage-ts-sdk/lib.esm/kv/client.js
import { StorageKv } from '../node/index.js';
import { KvIterator } from './iterator.js';
import { MAX_QUERY_SIZE } from './constants.js';
export class KvClient {
    inner;
    constructor(rpc) {
        const client = new StorageKv(rpc);
        this.inner = client;
    }
    newIterator(streamId, version) {
        return new KvIterator(this, streamId, version);
    }
    async getValue(streamId, key, version) {
        let val = {
            data: '',
            size: 0,
            version: version || 0,
        };
        while (true) {
            const seg = await this.inner.getValue(streamId, key, val.data.length, MAX_QUERY_SIZE, version);
            if (seg === undefined) {
                return null;
            }
            if (val.version === Number.MAX_SAFE_INTEGER) {
                val.version = seg.version;
            }
            else if (val.version !== seg.version) {
                val.version = seg.version;
                val.data = '';
            }
            val.size = seg.size;
            const segData = Buffer.from(seg.data, 'base64');
            const valData = Buffer.from(val.data, 'base64');
            val.data = Buffer.concat([valData, segData]).toString('base64');
            if (seg.size == segData.length + valData.length) {
                return val;
            }
        }
    }
    async get(streamId, key, startIndex, length, version) {
        return this.inner.getValue(streamId, key, startIndex, length, version);
    }
    async getNext(streamId, key, startIndex, length, inclusive, version) {
        return this.inner.getNext(streamId, key, startIndex, length, inclusive, version);
    }
    /**
     * Like `getNext`, but always returns the *fully assembled* value:
     * tries a single-RPC `getNext` with `MAX_QUERY_SIZE` budget first
     * (covers the common case where the value fits in one chunk), and
     * falls back to `getValue`'s chunk-assembly loop if the value
     * spans multiple chunks. Callers don't need to think about
     * `startIndex`/`length` or partial-value reassembly.
     *
     * Returns `null` at end-of-stream (when the underlying RPC
     * returns null/undefined for "no key found").
     */
    async getNextWithValue(streamId, key, inclusive, version) {
        const seg = await this.inner.getNext(streamId, key, 0, MAX_QUERY_SIZE, inclusive, version);
        if (!seg)
            return null;
        const segData = Buffer.from(seg.data, 'base64');
        if (segData.length === seg.size) {
            // Fast path: the entire value fit in the first chunk. 1 RPC total.
            return seg;
        }
        // Slow path: value exceeds MAX_QUERY_SIZE; fetch the rest.
        // The response's `seg.key` is base64-encoded; decode to raw
        // bytes before re-passing so the inner getValue's keyToBase64
        // doesn't try to arrayify a base64 string.
        const keyBytes = Buffer.from(seg.key, 'base64');
        const full = await this.getValue(streamId, keyBytes, version);
        if (full === null)
            return null;
        return {
            key: seg.key,
            data: full.data,
            size: full.size,
            version: seg.version,
        };
    }
    async getPrev(streamId, key, startIndex, length, inclusive, version) {
        return this.inner.getPrev(streamId, key, startIndex, length, inclusive, version);
    }
    async getFirst(streamId, startIndex, length, version) {
        return this.inner.getFirst(streamId, startIndex, length, version);
    }
    async getLast(streamId, startIndex, length, version) {
        return this.inner.getLast(streamId, startIndex, length, version);
    }
    async getTransactionResult(txSeq) {
        return this.inner.getTransactionResult(txSeq);
    }
    async getHoldingStreamIds() {
        return this.inner.getHoldingStreamIds();
    }
    async hasWritePermission(account, streamId, key, version) {
        return this.inner.hasWritePermission(account, streamId, key, version);
    }
    async isAdmin(account, streamId, version) {
        return this.inner.isAdmin(account, streamId, version);
    }
    async isSpecialKey(streamId, key, version) {
        return this.inner.isSpecialKey(streamId, key, version);
    }
    async isWriterOfKey(account, streamId, key, version) {
        return this.inner.isWriterOfKey(account, streamId, key, version);
    }
    async isWriterOfStream(account, streamId, version) {
        return this.inner.isWriterOfStream(account, streamId, version);
    }
}
